core/src/appleMain/kotlin/CBPeripheralCoreBluetoothPeripheral.kt

```diff
@@ -393,9 +402,42 @@ internal class CBPeripheralCoreBluetoothPeripheral(
         }
 
         val platformDescriptor = discoveredServices.obtain(descriptor)
-        return connection.execute<DidUpdateValueForDescriptor> {
+        val updatedDescriptor = connection.execute<DidUpdateValueForDescriptor> {
             centralManager.read(cbPeripheral, platformDescriptor)
-        }.descriptor.value as NSData
+        }.descriptor
+
+        // CBDescriptor "value" property type depends on cbDescriptor.UUID.UUIDString
+        // https://developer.apple.com/documentation/corebluetooth/characteristic-descriptors
+        // see type conversion table https://github.com/JuulLabs/kable/pull/706
+        return when (val value = updatedDescriptor.value) {
+            is NSData -> value
+            is NSString -> value.dataUsingEncoding(NSUTF8StringEncoding) ?: byteArrayOf().toNSData().also {
+                logger.warn {
+                    message = "Failed to decode descriptor"
+                    detail("type", "NSString")
+                    detail(descriptor)
+                }
+            }
+            is NSNumber -> {
+                when (updatedDescriptor.UUID.UUIDString) {
+                    CBUUIDCharacteristicExtendedPropertiesString,
+                    CBUUIDClientCharacteristicConfigurationString,
+                    CBUUIDServerCharacteristicConfigurationString,
+                    CBUUIDL2CAPPSMCharacteristicString,
+                    -> value.unsignedShortValue.toNSData()
```

Comment:
This feels like a good approach to me. Thanks for the thorough explanation around this. ❤️